### PR TITLE
[Orchestrator] Remove validation failure on timeout

### PR DESCRIPTION
**Pull request by loic-sharma**
Change the Orchestrator so that:

* Validation timeouts no longer fail validations
* An email is sent when a validation set reaches the `SendPackageIsTakingTooLongToValidateMessageAfter` threshold. This is currently set to 50m due to the state of our current pipeline.
* Metrics are emitted when validators reach their `TimeoutAfter` threshold
* A validation set no longer gets processed after its duration passes the `TimeoutValidationSetAfter` threshold

An alert will be raised when a validation set times out. The on-call engineer will need to either:
1. Revalidate the package
2. Mark the package as invalid in the Gallery (PackageStatusKey = Invalid) and send a canned email to the user. We should consider making an admin panel for this.

Total validation times since March 1st:

50% | 90% | 95% | 99%
--- | --- | --- | ---
2.2m | 11m | 47m | 4.7h

Progress on https://github.com/NuGet/Engineering/issues/1144

### CODE REVIEW

File: src/NuGet.Services.Validation.Orchestrator/Configuration/ValidationConfiguration.cs [hunk 1]
```diff
@@ -39,5 +39,15 @@ public class ValidationConfiguration
...
+
+        /// <summary>
+        /// The threshold until which an email will be sent out due to a validation set taking too long.
+        /// </summary>
+        public TimeSpan SendPackageIsTakingTooLongToValidateMessageAfter { get; set; }
+
+        /// <summary>
+        /// The threshold until a validation set is no longer processed.
+        /// </summary>
+        public TimeSpan TimeoutValidationSetAfter { get; set; }
```
Comment (chenriksson) [inline src/NuGet.Services.Validation.Orchestrator/Configuration/ValidationConfiguration.cs]: `SendPackageIsTakingTooLongToValidateMessageAfter` is so long... is there a good, short name?

Maybe `ValidationSetNotificationTimeout` and `ValidationSetProcessorTimeout`?

File: src/NuGet.Services.Validation.Orchestrator/Telemetry/TelemetryService.cs [hunk 2]
```diff
@@ -15,6 +15,8 @@ public class TelemetryService : ITelemetryService
         private const string DurationToValidationSetCreationSeconds = Prefix + "DurationToValidationSetCreationSeconds";
         private const string PackageStatusChange = Prefix + "PackageStatusChange";
         private const string TotalValidationDurationSeconds = Prefix + "TotalValidationDurationSeconds";
+        private const string SentValidationTakingTooLongMessage = Prefix + "SentValidationTakingTooLongMessage";
+        private const string ValidationSetTimeout = Prefix + "TotalValidationDurationSeconds";
```
Comment (chenriksson) [inline src/NuGet.Services.Validation.Orchestrator/Telemetry/TelemetryService.cs]: Do we need a `validationSetTimeout`? Shouldn't it be equal to the max timeout of all validators/processors?

File: src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs [hunk 3]
```diff
@@ -154,13 +150,63 @@ ValidationConfigurationItem GetValidationConfigurationItem(string validationName
...
             else
             {
-                // No failed validations and some validations are still in progress.
-                // Scheduling another check
-                var messageData = new PackageValidationMessageData(package.PackageRegistration.Id, package.Version, validationSet.ValidationTrackingId);
-                await _validationEnqueuer.StartValidationAsync(messageData, DateTimeOffset.UtcNow + _validationConfiguration.ValidationMessageRecheckPeriod);
+                // There are no failed validations and some validations are still in progress. Update
+                // the validation set's Updated field and send a notice if the validation set has reached
+                // the configured threshold.
+                var perviousUpdateTime = validationSet.Updated;
```
Comment (chenriksson) [inline src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs]: haha... nice typo!

File: src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs [hunk 4]
```diff
@@ -154,13 +150,63 @@ ValidationConfigurationItem GetValidationConfigurationItem(string validationName
...
+                if (validationSetDuration <= _validationConfiguration.TimeoutValidationSetAfter)
+                {
+                    var messageData = new PackageValidationMessageData(package.PackageRegistration.Id, package.Version, validationSet.ValidationTrackingId);
+                    var postponeUntil = DateTimeOffset.UtcNow + _validationConfiguration.ValidationMessageRecheckPeriod;
+
+                    await _validationEnqueuer.StartValidationAsync(messageData, postponeUntil);
+                }
+                else
+                {
+                    _telemetryService.TrackValidationSetTimeout(package.PackageRegistration.Id, package.NormalizedVersion, validationSet.ValidationTrackingId.ToString());
```
Comment (chenriksson) [inline src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs]: Is there any overhead that would contribute to validation set latency, other than validator/processor latencies? Just wondering why we need this timeout?

File: src/NuGet.Services.Validation.Orchestrator/Telemetry/TelemetryService.cs (hunk 2 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/Telemetry/TelemetryService.cs]: Good question. The validators' timeouts are alertable thresholds: something has gone wrong and validation is taking too long, but, it's possible that the validation may still succeed. On the other hand, the `validationSetTimeout` is a fail-safe to ensure the Orchestrator does not process "dead" validation sets forever.

TLDR: Ideally, `ValidationSetTimeout` is a value that is much higher than all of the validators' `timeoutAfter`. This isn't true today, but after we've improved the virus scanner's latency we should be able to bring down validator's `timeoutAfter`.

File: src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs (hunk 4 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs]: See my other comment about the differences between the validation set's timeout and the individual validators' timeout.

Also, the Orchestrator should be resilient to Validators misbehaving (either due to bugs or misconfigurations). That alone requires that the Orchestrator has its own fail-safe config :)

File: src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs (hunk 3 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs]: Oops!

File: src/NuGet.Services.Validation.Orchestrator/Configuration/ValidationConfigurationItem.cs [hunk 5]
```diff
@@ -17,9 +17,9 @@ public class ValidationConfigurationItem
         public string Name { get; set; }
 
         /// <summary>
-        /// Timeout after which started validation is considered failed if it didn't produce any outcome
+        /// Timeout after which a validation will no longer be processed if it hasn't produced an outcome.
         /// </summary>
-        public TimeSpan FailAfter { get; set; }
+        public TimeSpan TimeoutAfter { get; set; }
```
Comment (agr) [inline src/NuGet.Services.Validation.Orchestrator/Configuration/ValidationConfigurationItem.cs]: Just `Timeout` maybe?

File: src/NuGet.Services.Validation.Orchestrator/MessageService.cs [hunk 6]
```diff
@@ -70,5 +70,13 @@ public void SendPackageSignedValidationFailedMessage(Package package)
             var galleryPackageUrl = string.Format(_emailConfiguration.PackageUrlTemplate, package.PackageRegistration.Id, package.NormalizedVersion);
             _coreMessageService.SendSignedPackageNotAllowedNotice(package, galleryPackageUrl, _emailConfiguration.AnnouncementsUrl, _emailConfiguration.TwitterUrl);
         }
+
+        public void SendPackageValidationTakingTooLongMessage(Package package)
+        {
+            package = package ?? throw new ArgumentNullException(nameof(package));
+
+            var galleryPackageUrl = string.Format(_emailConfiguration.PackageUrlTemplate, package.PackageRegistration.Id, package.NormalizedVersion);
```
Comment (agr) [inline src/NuGet.Services.Validation.Orchestrator/MessageService.cs]: I have a feeling it is time to introduce small helper method that given the `packageId` would return package URL. Don't want that `string.Format` line copy-pasted any more.

File: src/NuGet.Services.Validation.Orchestrator/Configuration/ValidationConfigurationItem.cs (hunk 5 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/Configuration/ValidationConfigurationItem.cs]: @agr and I settled on `TrackAfter` as it is less confusing.

File: src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs [hunk 7]
```diff
@@ -44,24 +48,16 @@ public class ValidationOutcomeProcessor : IValidationOutcomeProcessor
             _messageService = messageService ?? throw new ArgumentNullException(nameof(messageService));
             _telemetryService = telemetryService ?? throw new ArgumentNullException(nameof(telemetryService));
             _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+
+            _validationConfigurationsByName = _validationConfiguration.Validations.ToDictionary(v => v.Name);
```
Comment (agr) [inline src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs]: Are we sure `_validationConfiguration.Validations` cannot be `null` here?

Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs]: This should be caught by the `ConfigurationValidator`'s `CheckValidationsNumber`.

---
In reply to: [174258183](https://github.com/NuGet/NuGet.Jobs/pull/368#discussion_r174258183) [](ancestors = 174258183)

File: src/NuGet.Services.Validation.Orchestrator/ValidationStorageService.cs [hunk 8]
```diff
@@ -90,6 +90,18 @@ public class ValidationStorageService : IValidationStorageService
             TrackValidationStatus(packageValidation);
         }
 
+        public Task UpdateValidationSetAsync(PackageValidationSet packageValidationSet)
+        {
+            packageValidationSet = packageValidationSet ?? throw new ArgumentNullException(nameof(packageValidationSet));
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/ValidationStorageService.cs]: >packageValidationSet [](start = 12, length = 20)

Needs to set the update timestamp. Validation set does have row version so we are good there.

File: src/NuGet.Services.Validation.Orchestrator/Telemetry/ITelemetryService.cs [hunk 9]
```diff
@@ -34,7 +34,17 @@ public interface ITelemetryService
         void TrackTotalValidationDuration(TimeSpan duration, bool isSuccess);
 
         /// <summary>
-        /// A counter metric emitted when a validator fails due to the <see cref="ValidationConfigurationItem.FailAfter"/>
+        /// A counter metric emitted when a notification is sent because a validation set takes too long.
+        /// </summary>
+        void TrackSentValidationTakingTooLongMessage(string packageId, string normalizedVersion, string validationTrackingId);
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/Telemetry/ITelemetryService.cs]: >validationTrackingId [](start = 104, length = 20)

Pass a GUID here and do the mapping below. This interface should not leak up type limitations IMO.

File: src/NuGet.Services.Validation.Orchestrator/ValidationStorageService.cs [hunk 10]
```diff
@@ -90,6 +90,18 @@ public class ValidationStorageService : IValidationStorageService
             TrackValidationStatus(packageValidation);
         }
 
+        public Task UpdateValidationSetAsync(PackageValidationSet packageValidationSet)
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/ValidationStorageService.cs]: This needs unit tests

File: src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs [hunk 11]
```diff
@@ -154,13 +150,63 @@ ValidationConfigurationItem GetValidationConfigurationItem(string validationName
...
+                var perviousUpdateTime = validationSet.Updated;
+                var validationSetDuration = DateTime.UtcNow - validationSet.Created;
+                var previousDuration = perviousUpdateTime - validationSet.Created;
+
+                await _validationStorageService.UpdateValidationSetAsync(validationSet);
+
+                if (validationSetDuration > _validationConfiguration.SendPackageIsTakingTooLongToValidateMessageAfter &&
+                    previousDuration <= _validationConfiguration.SendPackageIsTakingTooLongToValidateMessageAfter)
+                {
+                    _messageService.SendPackageValidationTakingTooLongMessage(package);
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs]: >SendPackageValidationTakingTooLongMessage [](start = 36, length = 41)

Let's try to only send this message on the first validation set. I don't think a revalidate should send out this email.

File: src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs [hunk 12]
```diff
@@ -154,13 +150,63 @@ ValidationConfigurationItem GetValidationConfigurationItem(string validationName
...
+                // There are no failed validations and some validations are still in progress. Update
+                // the validation set's Updated field and send a notice if the validation set has reached
+                // the configured threshold.
+                var perviousUpdateTime = validationSet.Updated;
+                var validationSetDuration = DateTime.UtcNow - validationSet.Created;
+                var previousDuration = perviousUpdateTime - validationSet.Created;
+
+                await _validationStorageService.UpdateValidationSetAsync(validationSet);
+
+                if (validationSetDuration > _validationConfiguration.SendPackageIsTakingTooLongToValidateMessageAfter &&
```
Comment (joelverhagen) [inline src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs]: >validationSetDuration [](start = 20, length = 21)

We should leave comment about how this, in addition to row version, ensures we send at most one "too long" email per validation set.

File: src/NuGet.Services.Validation.Orchestrator/Configuration/ValidationConfiguration.cs (hunk 1 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/Configuration/ValidationConfiguration.cs]: Fixed

File: src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs (hunk 3 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs]: Fixed

File: src/NuGet.Services.Validation.Orchestrator/Configuration/ValidationConfigurationItem.cs (hunk 5 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/Configuration/ValidationConfigurationItem.cs]: Fixed

File: src/NuGet.Services.Validation.Orchestrator/ValidationStorageService.cs (hunk 8 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/ValidationStorageService.cs]: Fixed

File: src/NuGet.Services.Validation.Orchestrator/Telemetry/ITelemetryService.cs (hunk 9 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/Telemetry/ITelemetryService.cs]: Fixed

File: src/NuGet.Services.Validation.Orchestrator/ValidationStorageService.cs (hunk 10 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/ValidationStorageService.cs]: Fixed

File: src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs (hunk 11 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs]: Fixed

File: src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs (hunk 12 above)
Comment (loic-sharma) [reply src/NuGet.Services.Validation.Orchestrator/ValidationOutcomeProcessor.cs]: Fixed

**Comment by loic-sharma** [comment]
:bell: All feedback addressed :)
